test/python/circuit/classical/test_expr_constructors.py

```diff
@@ -168,6 +204,31 @@ def test_binary_functions_lift_scalars(self, function, left, right):
         self.assertEqual(function(left, right), function(left, expr.lift(right)))
         self.assertEqual(function(left, right), function(expr.lift(left), expr.lift(right)))
 
+    @ddt.data(
+        (expr.bit_and, 6, 7),
+        (expr.bit_or, 5, 6),
+        (expr.bit_xor, 255, 254),
+        (expr.equal, 254, 255),
+        (expr.not_equal, 255, 255),
+        (expr.less, 5, 4),
+        (expr.less_equal, 3, 3),
+        (expr.greater, 254, 255),
+        (expr.greater_equal, 4, 5),
+    )
+    @ddt.unpack
+    def test_binary_functions_lift_scalars_const(self, function, left, right):
+        """If one operand is an expr with a const type, the other scalar should be lifted as const.
+        Note that logical operators (e.g. logic_and, logic_or) are excluded since these lift operands
+        independently."""
+        self.assertEqual(
+            function(expr.lift(left, try_const=True), right),
+            function(expr.lift(left, try_const=True), expr.lift(right, try_const=True)),
+        )
+        self.assertEqual(
+            function(left, expr.lift(right, try_const=True)),
+            function(expr.lift(left, try_const=True), expr.lift(right, try_const=True)),
+        )
```

Comment:
I think perhaps this is indicating missing logic in `logic_and` and `logic_or` - it always lifts to `Bool(const=False)`, but `logic_and(False, True)` should have type `Bool(const=True)`, as should `logic_and(False, expr.Var.new("a", types.Bool(const=True))`.  `logic_not` does the same thing, I think - `logic_not(False)` should be `Bool(const=True)`.

Reply:
You're right, and that's actually a bug I think from when I switched the default of `lift`'s `try_const` to `False`.

Reply:
Hmm, actually, this was intended. It's supposed to be consistent with `lift` going for non-const by default. Since the operands are lifted independently, this is the consistent behavior. We could instead do what the other operations do, which is to use `_lift_binary_operands` before ultimately coercing both the `bool`. But this would also widen integers to the same width, which isn't the current behavior. Thoughts?

Reply:
> the operands are lifted independently

This is just an implementation detail of the pre-`const` type system that there was no influence of one binary operand on the other - I don't think there's any indication that it's an API contract of the inference?  I'd expect the inference signature of `&&` to be `const bool * const bool -> const bool` as a user, and similarly for `! : const bool -> const bool`.

If you're concerned about the consistency of `lift`, I think it might be indicative that we might want to get rid of `try_const`, or default it to `True`?

I think the QPY concern could in theory be handled separately (I agree it's a valid one), such as threading a keyword arg through `dump` like `allow_lossy_expr_types` or something?

Reply:
This is also all resolved with the change to `Expr.const`